### PR TITLE
feat(policies): add policy and group references to policy evaluations

### LINKED ISSUE #1405: Add group name to policy evaluations
When a policy is included in a policy group, this context is lost when the attestation is sent. Adding a `group` field in the evaluation result would solve this.

### PR DESCRIPTION
**Pull request by jiparis**
This PR supersedes https://github.com/chainloop-dev/chainloop/pull/1444. 

Policies evaluated in the context of a group will have two new `policy_reference` and `group_reference` properties in the evaluation result with a full `intoto.ResourceDescriptor` for the group and the policy itself.

This is example taken with `chainloop wf run describe --digest xxxxx -o statement`, with a group which defines a `sbom` material and several policy evaluations:
```json
...
              "annotations": {
                  "category": "sbom"
               },
               "description": "Checks that the SBOM is not older than a specified threshold. Supports CycloneDX.\nIt accepts a limit argument denoting the number of days, 30 by default:\n  with:\n    limit: 20\n",
               "group_reference": {
                  "annotations": {
                     "name": "sbom-quality",
                     "organization": ""
                  },
                  "digest": {
                     "sha256": "b7b8fac912c7c2f404ec0025daaa40b82f30bb91567660e6f8e7dc8a0327e67e"
                  },
                  "name": "sbom-quality",
                  "uri": "file://test/policy-group.yaml"
               },
               "material_name": "sbom",
               "name": "sbom-freshness",
               "policy_reference": {
                  "annotations": {
                     "name": "sbom-freshness",
                     "organization": ""
                  },
                  "digest": {
                     "sha256": "f6848d595e910686669a998efb1911f6e001efcac9fc3eefeb0b1b8f3ffd792c"
                  },
                  "name": "sbom-freshness",
                  "uri": "chainloop://localhost:8002/sbom-freshness"
               },
               "skipped": false,
               "type": "SBOM_CYCLONEDX_JSON",
               "violations": [
                  {
                     "message": "SBOM created at: 2024-07-23T15:58:37+00:00 which is too old (freshness limit set to 30 days)",
                     "subject": "sbom-freshness"
                  }
               ],
               "with": {
                  "freshness": "20"
               }
...
```
This is another example with a custom policy (note the `organization` annotation):
```json
            {
               "annotations": {
                  "category": ""
               },
               "description": "This is my policy",
               "material_name": "sbom",
               "name": "my-policy",
               "policy_reference": {
                  "annotations": {
                     "name": "my-policy",
                     "organization": "my-org"
                  },
                  "digest": {
                     "sha256": "d56ff0d5dacb322e4bf05380687a2e2e90193ad5420da927e2cc0cbf95855d83"
                  },
                  "name": "my-policy",
                  "uri": "chainloop://localhost:8002/my-policy?org=my-org"
               },
               "skip_reasons": [
                  "this is a test"
               ],
               "skipped": true,
               "type": "SBOM_CYCLONEDX_JSON"
            }
```

Also note that this PR uses correctly the intoto descriptor fields, in particular `name` and `uri` now contains the policy/group name and it's URI (previously the `name` field contained the URI)

Closes #1405

### CODE REVIEW

**Review by javirln — APPROVED** [review]

File: pkg/attestation/crafter/api/attestation/v1/crafting_state.proto [hunk 1]
```diff
@@ -146,6 +146,9 @@ message PolicyEvaluation {
   // Evaluation messages, intended to communicate evaluation errors (invalid input)
   repeated string skip_reasons = 14;
 
+  // Group this evaluated policy belongs to, if any
+  string group_name = 15;
```
Comment (migmartri) [inline pkg/attestation/crafter/api/attestation/v1/crafting_state.proto:150]: I thought we were going to add some sort of group-ref?

Comment (jiparis) [reply pkg/attestation/crafter/api/attestation/v1/crafting_state.proto:150]: Right. I'll set this on draft.

Comment (jiparis) [reply pkg/attestation/crafter/api/attestation/v1/crafting_state.proto]: Done, please review again

File: pkg/attestation/crafter/api/attestation/v1/crafting_state.proto [hunk 2]
```diff
@@ -146,10 +147,21 @@ message PolicyEvaluation {
...
   message Violation {
     string subject = 1 [(buf.validate.field).required = true];
     string message = 2 [(buf.validate.field).required = true];
   }
+
+  message Reference {
+    string name = 1;
+    string digest = 2;
+    string uri = 3;
+    string org_name = 4;
```
Comment (migmartri) [inline pkg/attestation/crafter/api/attestation/v1/crafting_state.proto:163]: you might want to add validation annotations here?

File: pkg/attestation/crafter/api/attestation/v1/crafting_state.proto [hunk 3]
```diff
@@ -146,10 +147,21 @@ message PolicyEvaluation {
   // Evaluation messages, intended to communicate evaluation errors (invalid input)
   repeated string skip_reasons = 14;
 
+  // Group this evaluated policy belongs to, if any
+  Reference policy_reference = 15;
```
Comment (migmartri) [inline pkg/attestation/crafter/api/attestation/v1/crafting_state.proto:151]: where are you setting the policy_reference value? I can't see any changes in the crafter pkg?

**Review by migmartri — APPROVED** [review]
See my comment re: not clear where the data is set

File: pkg/attestation/crafter/api/attestation/v1/crafting_state.proto (hunk 3 above)
Comment (jiparis) [reply pkg/attestation/crafter/api/attestation/v1/crafting_state.proto:151]: This is set during policy evaluation, which happens in the crafting stage, see `policies.go` below.

Comment (migmartri) [reply pkg/attestation/crafter/api/attestation/v1/crafting_state.proto:151]: oh I see, thanks

**Pull request merged**